src/coreComponents/dataRepository/wrapperHelpers.hpp

```diff
@@ -490,6 +490,114 @@ pullDataFromConduitNode( Array< T, NDIM, PERMUTATION > & var,
+
+  // ArrayOfArrays::m_offsets
+  INDEX_TYPE const * const offsets = var.getOffsets();
+  conduit::DataType const offsetsType( conduitTypeInfo< INDEX_TYPE >::id, numArrays+1 );
+  node[ "__offsets__" ].set_external( offsetsType, const_cast< void * >( static_cast< void const * >( offsets ) ) );
+
+  // ArrayOfArrays::m_sizes
+  INDEX_TYPE const * const sizes = var.getSizes();
+  conduit::DataType const sizesType( conduitTypeInfo< INDEX_TYPE >::id, numArrays );
+  node[ "__sizes__" ].set_external( sizesType, const_cast< void * >( static_cast< void const * >( sizes ) ) );
+
+  // Push the data into conduit
+  constexpr int conduitTypeID = conduitTypeInfo< T >::id;
+  constexpr int sizeofConduitType = conduitTypeInfo< T >::sizeOfConduitType;
+  conduit::DataType const dtype( conduitTypeID, offsets[numArrays] * sizeof( T ) / sizeofConduitType );
+  void * const ptr = const_cast< void * >( static_cast< void const * >( var.getValues() ) );
+  node[ "__values__" ].set_external( dtype, ptr );
+}
+
+template< typename T, typename INDEX_TYPE >
+std::enable_if_t< bufferOps::can_memcpy< T > >
+pullDataFromConduitNode( ArrayOfArrays< T, INDEX_TYPE > & var,
+                         conduit::Node const & node )
+{
+
+  // numArrays node
+  conduit::Node const & numArraysNode = node.fetch_existing( "__numberOfArrays__" );
+  INDEX_TYPE const * const numArrays = numArraysNode.value();
+
+  // offsets node
+  conduit::Node const & offsetsNode = node.fetch_existing( "__offsets__" );
+  conduit::DataType const & offsetsDataType = offsetsNode.dtype();
+  INDEX_TYPE const * const offsets = offsetsNode.value();
+  INDEX_TYPE const sizeOffsets = offsetsDataType.number_of_elements();
+
+  // sizes node
+  conduit::Node const & sizesNode = node.fetch_existing( "__sizes__" );
+  conduit::DataType const & sizesDataType = sizesNode.dtype();
+  INDEX_TYPE const * const sizes = sizesNode.value();
+  INDEX_TYPE const sizeSizes = sizesDataType.number_of_elements();
+
+  // Check that the numArrays, sizes and offsets are consistent.
+  GEOS_ERROR_IF_NE( *numArrays, sizeSizes );
+  GEOS_ERROR_IF_NE( *numArrays+1, sizeOffsets );
+
+  // values node
+  conduit::Node const & valuesNode = node.fetch_existing( "__values__" );
+  conduit::DataType const & valuesDataType = valuesNode.dtype();
+  const INDEX_TYPE valuesSize = valuesDataType.number_of_elements();
+
+  // should preallocate var.m_values with estimated sizes
+  INDEX_TYPE const arraySizeEstimate = (*numArrays)==0 ? 0 : valuesSize / (*numArrays);
+  var.resize( *numArrays, arraySizeEstimate );
+  var.reserveValues( valuesSize );
+
+  // correctly set the sizes and capacities of each sub-array
+  localIndex allocatedSize = 0;
+  for( INDEX_TYPE i = 0; i < *numArrays; ++i )
+  {
+    INDEX_TYPE const arrayAllocation = offsets[i+1] - offsets[i];
+    var.setCapacityOfArray( i, arrayAllocation );
+    var.resizeArray( i, sizes[ i ] );
+    allocatedSize += arrayAllocation;
+  }
```

Comment:
@corbett5
I have been trying a few iterations with this section of code. I can't seem to get it right. Can you have a close look at the attempt to pre-allocate values and the subsequent capacity/resize of each array to make sure it makes sense?

Reply:
For this whole function I think I would do this

```c++
var.resizeFromOffsets( *numArrays, offsets );

for (INDEX_TYPE i = 0; i < numArrays; ++i )
{
  T const * begin = valuesNode.data_ptr() + offsets[i];
  T const * end = begin + sizes[ i ];
  var.appendToArray( i, begin, end );
}
```

Reply:
@corbett5 Is the name `resizeFromOffsets` accurate? Looking that the function should it be `reserveFromOffsets`?